x/stakeibc/keeper/unbonding_records.go

```diff
@@ -235,6 +245,111 @@ func (k Keeper) GetUnbondingICAMessages(
+
+	// Store the delegation of each validator that was expected *after* the originally
+	// planned unbonding went through
+	// e.g. If the validator had 10 before unbonding, and in the first pass, 3 was
+	//      supposed to be unbonded, their delegation after the first pass is 7
+	totalRemainingDelegationsAcrossBatch := sdk.ZeroDec()
+	remainingDelegationsInBatchByVal := map[string]sdk.Dec{}
+	for _, capacity := range unbondCapacities {
+		// Only add validators that were in the initial unbonding plan
+		// The delegation after the first pass is calculated by taking the "current delegation"
+		// (aka delegation before unbonding) and subtracting the unbond amount
+		if initialUnbondAmount, ok := initialUnbondAmountFromBatchByVal[capacity.ValidatorAddress]; ok {
+			remainingDelegation := sdk.NewDecFromInt(capacity.CurrentDelegation.Sub(initialUnbondAmount))
+
+			remainingDelegationsInBatchByVal[capacity.ValidatorAddress] = remainingDelegation
+			totalRemainingDelegationsAcrossBatch = totalRemainingDelegationsAcrossBatch.Add(remainingDelegation)
+		}
+	}
+
+	// This is to protect against a division by zero error, but this would technically be possible
+	// if the 32 validators with the most capacity were all 0 weight
+	if totalRemainingDelegationsAcrossBatch.IsZero() {
+		return finalUnbondings, errors.New("no delegations to redistribute during consolidation")
+	}
+
+	// Before we start dividing up the excess, make sure we have sufficient stake in the capped set to cover it
+	if sdk.NewDecFromInt(totalExcessAmount).GT(totalRemainingDelegationsAcrossBatch) {
+		return finalUnbondings, errors.New("not enough exisiting delegation in the batch to cover the excess")
+	}
+
+	// Loop through the original unbonding messages and proportionally divide out
+	// the excess amongst the validators in the set
+	excessRemaining := totalExcessAmount
+	for i := range unbondingsBatch {
+		unbonding := unbondingsBatch[i]
+
+		var validatorUnbondIncrease sdkmath.Int
+		if i != len(unbondingsBatch)-1 {
+			// For all but the last validator, calculate their unbonding increase by
+			// splitting the excess proportionally in line with their remaining delegation
+			remainingDelegation, ok := remainingDelegationsInBatchByVal[unbonding.Validator]
+			if !ok {
+				return finalUnbondings, fmt.Errorf("validator %s not found in initial unbonding plan", unbonding.Validator)
+			}
+			unbondIncreaseProportion := remainingDelegation.Quo(totalRemainingDelegationsAcrossBatch)
+			validatorUnbondIncrease = sdk.NewDecFromInt(totalExcessAmount).Mul(unbondIncreaseProportion).TruncateInt()
+
+			// Decrement excess
+			excessRemaining = excessRemaining.Sub(validatorUnbondIncrease)
+		} else {
+			// The last validator in the set should get any remainder from int truction
+			// First confirm the validator has sufficient remaining delegation to cover this
+			remainingDelegation, ok := remainingDelegationsInBatchByVal[unbonding.Validator]
```

Comment:
nit: could we move this above the `if i != len(unbondingsBatch) - 1` statement, to avoid duplicating code across the `if` and `else`?

Reply:
yes, great call!

Reply:
done [[here](https://github.com/Stride-Labs/stride/pull/1018/commits/fb122c549a7df2ef85fdbbb8b2798bfb3f9b8fd6)]

Reply:
awesome, thank you!